File: august-24/august-3.js
/*

Given the head of a singly linked list, swap every two nodes and return its head.

For example, given 1 -> 2 -> 3 -> 4, return 2 -> 1 -> 4 -> 3.

*/

class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

function swapPairs(head) {
    if (!head || !head.next) return head;

    let dummy = new ListNode(0);
    dummy.next = head;
    let prev = dummy;

    while (head && head.next) {
        let firstNode = head;
        let secondNode = head.next;

        // Swapping
        firstNode.next = secondNode.next;
        secondNode.next = firstNode;
        prev.next = secondNode;

        // Reinitializing the pointers for the next swap
        prev = firstNode;
        head = firstNode.next; // move head to the next pair
    }

    return dummy.next;
}

// Helper function to create linked list from array
function createLinkedList(arr) {
    let dummy = new ListNode();
    let current = dummy;
    for (let val of arr) {
        current.next = new ListNode(val);
        current = current.next;
    }
    return dummy.next;
}

// Helper function to print linked list
function printLinkedList(head) {
    let current = head;
    while (current) {
        process.stdout.write(current.val + " -> ");
        current = current.next;
    }
    console.log("null");
}

// Test Case:
let head = createLinkedList([1, 2, 3, 4]);
console.log("Original list:");
printLinkedList(head);
head = swapPairs(head);
console.log("Swapped list:");
printLinkedList(head);

File: august-24/august-3.py
# Given the head of a singly linked list, swap every two nodes and return its head.

# For example, given 1 -> 2 -> 3 -> 4, return 2 -> 1 -> 4 -> 3.

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def swapPairs(head):
    # Create a dummy node to handle edge cases
    dummy = ListNode(0)
    dummy.next = head
    prev = dummy

    # Traverse the list while there are pairs to swap
    while prev.next and prev.next.next:
        first = prev.next
        second = prev.next.next

        # Perform the swap
        first.next = second.next
        second.next = first
        prev.next = second

        # Move to the next pair
        prev = first

    return dummy.next

# Helper function to print the list
def printList(head):
    current = head
    while current:
        print(current.val, end=" -> ")
        current = current.next
    print("None")

# Example usage
head = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
print("Original list:")
printList(head)

swapped_head = swapPairs(head)
print("Swapped list:")
printList(swapped_head)
